### PR TITLE
Geocoder plugin: support other built-in providers

### PR DESCRIPTION
**Pull request by prusswan**
tested with Nominatim, Photon, Google:
```Python
from branca.element import MacroElement
from jinja2 import Template

from folium.elements import JSCSSMixin
from folium.utilities import parse_options

import folium

class EnhancedGeocoder(JSCSSMixin, MacroElement):
    """A simple geocoder for Leaflet that by default uses OSM/Nominatim.

    Please respect the Nominatim usage policy:
    https://operations.osmfoundation.org/policies/nominatim/
    Parameters
    ----------
    collapsed: bool, default False
        If True, collapses the search box unless hovered/clicked.
    position: str, default 'topright'
        Choose from 'topleft', 'topright', 'bottomleft' or 'bottomright'.
    add_marker: bool, default True
        If True, adds a marker on the found location.
    geocode_zoom: int, default 11
        Set zoom level used for displaying the geocode result, note that this only has an effect when add_marker is set to False. Set this to None to preserve the current map zoom level.
    geocode_provider: str, default 'nominatim'
        Defaults to "nominatim", see https://github.com/perliedman/leaflet-control-geocoder/tree/2.4.0/src/geocoders for other built-in providers.
    geocode_provider_options: dict, default {}
        For use with specific providers that may require api keys or other parameters.

    For all options see https://github.com/perliedman/leaflet-control-geocoder

    """

    _template = Template(
        """
        {% macro script(this, kwargs) %}

            var geocoderOpts_{{ this.get_name() }} = {{ this.options|tojson }};

            // note: geocoder name should start with lowercase
            var geocoderName_{{ this.get_name() }} = geocoderOpts_{{ this.get_name() }}["geocodeProvider"];

            var customGeocoder_{{ this.get_name() }} = L.Control.Geocoder[ geocoderName_{{ this.get_name() }} ](
                geocoderOpts_{{ this.get_name() }}['geocodeProviderOptions']
            );
            geocoderOpts_{{ this.get_name() }}["geocoder"] = customGeocoder_{{ this.get_name() }};

            L.Control.geocoder(
                geocoderOpts_{{ this.get_name() }}
            ).on('markgeocode', function(e) {
                var zoom = geocoderOpts_{{ this.get_name() }}['geocodeZoom'] || {{ this._parent.get_name() }}.getZoom();
                {{ this._parent.get_name() }}.setView(e.geocode.center, zoom);
            }).addTo({{ this._parent.get_name() }});

        {% endmacro %}
    """
    )

    default_js = [
        (
            "Control.Geocoder.js",
            "https://unpkg.com/leaflet-control-geocoder/dist/Control.Geocoder.js",
        )
    ]
    default_css = [
        (
            "Control.Geocoder.css",
            "https://unpkg.com/leaflet-control-geocoder/dist/Control.Geocoder.css",
        )
    ]

    def __init__(
        self,
        collapsed: bool = False,
        position: str = "topright",
        add_marker: bool = True,
        geocode_zoom: int | None = 11,
        geocode_provider: str = "nominatim",
        geocode_provider_options: dict = {},
        **kwargs
    ):
        super().__init__()
        self._name = "Geocoder"
        self.options = parse_options(
            collapsed=collapsed,
            position=position,
            default_mark_geocode=add_marker,
            geocode_zoom=geocode_zoom,
            geocode_provider=geocode_provider,
            geocode_provider_options=geocode_provider_options,
            **kwargs
        )

google_key = "placeholder"
gc2_default = EnhancedGeocoder() # nominatim
#gc2_photon = EnhancedGeocoder(geocode_provider="photon")
#gc2_google = EnhancedGeocoder(geocode_provider="google",
#                           geocode_provider_options={"apiKey": google_key})


m2 = folium.Map(location=[43.7798, 11.24148], zoom_start=13, tiles="cartodb positron")

gc2_default.add_to(m2)
#gc2_photon.add_to(m2)
#gc2_google.add_to(m2)

folium.LayerControl().add_to(m2)
m2
```

### CODE REVIEW

File: folium/plugins/geocoder.py [hunk 1]
```diff
@@ -19,6 +19,16 @@ class Geocoder(JSCSSMixin, MacroElement):
         Choose from 'topleft', 'topright', 'bottomleft' or 'bottomright'.
     add_marker: bool, default True
         If True, adds a marker on the found location.
+    geocodeProvider: str, default 'nominatim'
```
Comment (Conengmo) [inline folium/plugins/geocoder.py:22]: please use snake_case in the Python code. The `parse_options` function turns them into camelCase.

File: folium/plugins/geocoder.py [hunk 2]
```diff
@@ -50,12 +155,24 @@ class Geocoder(JSCSSMixin, MacroElement):
         )
     ]
 
-    def __init__(self, collapsed=False, position="topright", add_marker=True, **kwargs):
+    def __init__(self, collapsed=False, position="topright", add_marker=True,
```
Comment (Conengmo) [inline folium/plugins/geocoder.py:158]: to satisify our linter these should all be on their own line. Please also add type hints. For example:

```
def __init__(
        self,
        collapsed: bool = False,
        position: str = "topright",
```

You can run the autoformatter/linter locally with precommit:

- Install precommit https://pre-commit.com/#install
- Enable precommit for this repo: `pre-commit install`
- Run precommit on all files: `pre-commit run --all-files`
  - or of course only on the files you touched

Comment (prusswan) [reply folium/plugins/geocoder.py]: Have applied the snake_case, type-hints and ran it through `pre-commit`. Please take a look when you have time.

**Review by Conengmo — CHANGES_REQUESTED** [review]
Looks good overall! Few small comments before we can merge this.

File: folium/plugins/geocoder.py [hunk 3]
```diff
@@ -50,12 +68,24 @@ class Geocoder(JSCSSMixin, MacroElement):
         )
     ]
 
-    def __init__(self, collapsed=False, position="topright", add_marker=True, **kwargs):
+    def __init__(
+        self,
+        collapsed: bool = False,
+        position: str = "topright",
+        add_marker: bool = True,
+        geocode_zoom: int | None = 11,
```
Comment (Conengmo) [inline folium/plugins/geocoder.py:76]: > as the default zoom level of 11 is not suitable for close-up locations (and only works when add_marker is False)

Question: is 11 a good default? Or would None be a better default?

And another question: should we check or correct this value when `add_marker` is False?

Comment (prusswan) [reply folium/plugins/geocoder.py]: 11 would preserve the current behavior so existing users would not be caught by surprise. The real gotcha would be the case when `add_marker`=True (when default handler in leaflet-control-geocoder will be using `fit_bounds` with no regard to current map zoom). What may not be obvious to normal folium users (with no knowledge of JS) is that both the default event handler for `markgeocode` as well as the one defined in `_template` would be triggered.

Not sure if there is a need to correct (beyond keeping the current default of 11). Can you explain more?

**Review by Conengmo — APPROVED** [review]

File: folium/plugins/geocoder.py (hunk 3 above)
Comment (Conengmo) [reply folium/plugins/geocoder.py]: > Not sure if there is a need to correct (beyond keeping the current default of 11).

That's good then. It was more a question for you. We can keep it like it is now then.

**Comment by Conengmo** [comment]
Thanks @prusswan! Nice addition!

**Pull request merged**